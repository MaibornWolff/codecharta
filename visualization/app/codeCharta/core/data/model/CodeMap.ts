export interface CodeMapNode {
    name: string;
    type: string;
    children?: CodeMapNode[];
    attributes: KeyValuePair;
    deltas?: KeyValuePair;
    link?: string;
    origin?: string;
    visible?: boolean;
    path?: string;
}

export interface CodeMap {
    fileName: string;
    apiVersion?: string;
    projectName: string;
    nodes: CodeMapNode;
    edges?: Edge[];
    attributeTypes?: AttributeTypes;
    blacklist?: Array<BlacklistItem>;
}

export interface Edge {
    fromNodeName: string;
    toNodeName: string;
    attributes: KeyValuePair;
    visible?: boolean;
}

export interface AttributeTypes {
    nodes?: {
        [key: string]: AttributeType
    };
    edges?: {
        [key: string]: AttributeType
    };
}

export enum AttributeType {
    absolute, relative
}

export interface BlacklistItem {
    path: string;
    type: BlacklistType;
}

export enum BlacklistType {
    hide = "hide",
    exclude = "exclude"
}

export interface KeyValuePair {
    [key: string]: number;
}