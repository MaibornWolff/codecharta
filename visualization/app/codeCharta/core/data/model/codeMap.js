export class CodeMap {

    constructor(fileName, projectName, root) {

        this.fileName = fileName;
        this.projectName = projectName;
        this.root = root;

    }


}