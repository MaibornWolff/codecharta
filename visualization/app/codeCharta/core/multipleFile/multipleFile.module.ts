"use strict"

import angular from "angular"
import { MultipleFileService } from "./multipleFile.service"
import "../../ui/dialog/dialog"

angular.module("app.codeCharta.core.multiple", ["app.codeCharta.ui.dialog"]).service(MultipleFileService.SELECTOR, MultipleFileService)
