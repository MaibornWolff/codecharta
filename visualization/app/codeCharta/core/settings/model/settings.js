/**
 * This class defines the available set of settings in the CodeCharta application
 */
export class Settings {

    /**
     * @constructor
     * @param {CodeMap} map
     * @param {Range} neutralColorRange
     * @param {string} areaMetric
     * @param {string} heightMetric
     * @param {string} colorMetric
     * @param {boolean} deltas
     * @param {Number} amountOfTopLabels
     * @param {Scale} scaling
     * @param {Scale} camera
     * @param {Number} margin
     */
    constructor(map, neutralColorRange, areaMetric, heightMetric, colorMetric, deltas, amountOfTopLabels, scaling, camera, margin) {

        /**
         * currently selected map
         * @type {CodeMap}
         */
        this.map = map;

        /**
         * currently set neutral color range. This defines the yellow color range and the whether higher or lower values are green or red.
         * @type {Range}
         */
        this.neutralColorRange = neutralColorRange;

        /**
         * currently selected area metric
         * @type {string}
         */
        this.areaMetric = areaMetric;

        /**
         * currently selected heightm metric
         * @type {string}
         */
        this.heightMetric = heightMetric;

        /**
         * currently selected color metric
         * @type {string}
         */
        this.colorMetric = colorMetric;

        /**
         * delta flag
         * @type {boolean}
         */
        this.deltas = deltas;

        /**
         * number of highest buildings with labels
         * @type {Number}
         */
        this.amountOfTopLabels = amountOfTopLabels;

        /**
         * scaling settings
         * @type {Scale}
         */
        this.scaling = scaling;

        /**
         * camera position
         * @type {Scale}
         */
        this.camera = camera;

        /**
         * margin between buildings
         * @type {number}
         */
        this.margin = margin;

    }

    /**
     * Imports the given settings values without replacing this object with the given one.
     * @param {Settings} settings given settings
     */
    importSettingValues(settings){
        this.map = settings.map;
        this.neutralColorRange = settings.neutralColorRange;
        this.areaMetric = settings.areaMetric;
        this.heightMetric = settings.heightMetric;
        this.colorMetric = settings.colorMetric;
        this.deltas = settings.deltas;
        this.amountOfTopLabels = settings.amountOfTopLabels;
        this.scaling = settings.scaling;
        this.camera = settings.camera;
        this.margin = settings.margin;
    }

}
