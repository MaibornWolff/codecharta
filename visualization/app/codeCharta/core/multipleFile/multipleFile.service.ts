import { AttributeType, CodeMap, CodeMapNode, Edge, BlacklistItem } from "../data/model/CodeMap"

export class MultipleFileService {
	public static SELECTOR = "multipleFileService"
	private projectNameArray = []
	private fileNameArray = []
	private edges: Edge[] = []
	private blacklist: BlacklistItem[] = []
	private attributeTypesEdge: { [key: string]: AttributeType } = {}
	private attributeTypesNode: { [key: string]: AttributeType } = {}

	public aggregateMaps(inputMaps: CodeMap[]): CodeMap {
		if (inputMaps.length == 1) {
			return inputMaps[0]
		}

		this.resetVariables()

		for (let inputMap of inputMaps) {
			this.projectNameArray.push(inputMap.projectName)
			this.fileNameArray.push(inputMap.fileName)
			this.setEdgesWithUpdatedPath(inputMap)
			this.setAttributeTypesByUniqueKey(inputMap)
			this.setBlacklistWithUpdatedPath(inputMap)
		}
		return this.getNewAggregatedMap(inputMaps)
	}

	private setEdgesWithUpdatedPath(inputMap) {
		if (!inputMap.edges) {
			return
		}

		for (let oldEdge of inputMap.edges) {
			let edge: Edge = {
				fromNodeName: this.getUpdatedPath(inputMap.fileName, oldEdge.fromNodeName),
				toNodeName: this.getUpdatedPath(inputMap.fileName, oldEdge.toNodeName),
				attributes: oldEdge.attributes,
				visible: oldEdge.visible
			}
			const equalEdgeItems = this.edges.filter(e => e.fromNodeName == edge.fromNodeName && e.toNodeName == edge.toNodeName)
			if (equalEdgeItems.length == 0) {
				this.edges.push(edge)
			}
		}
	}

	private setBlacklistWithUpdatedPath(inputMap) {
		if (!inputMap.blacklist) {
			return
		}

		for (let oldBlacklistItem of inputMap.blacklist) {
			let blacklistItem: BlacklistItem = {
				path: this.getUpdatedBlacklistItemPath(inputMap.fileName, oldBlacklistItem.path),
				type: oldBlacklistItem.type
			}
			const equalBlacklistItems = this.blacklist.filter(b => b.path == blacklistItem.path && b.type == blacklistItem.type)
			if (equalBlacklistItems.length == 0) {
				this.blacklist.push(blacklistItem)
			}
		}
	}

	private getUpdatedBlacklistItemPath(fileName, path) {
		if (this.isAbsoluteRootPath(path)) {
			return this.getUpdatedPath(fileName, path)
		}
		return path
	}

	private isAbsoluteRootPath(path: string) {
		return path.startsWith("/root/")
	}

	private setAttributeTypesByUniqueKey(inputMap) {
		const types = inputMap.attributeTypes
		if (types && types.nodes) {
			for (let key in types.nodes) {
				this.attributeTypesNode[key] = types.nodes[key]
			}
		}
		if (types && types.edges) {
			for (let key in types.edges) {
				this.attributeTypesEdge[key] = types.edges[key]
			}
		}
	}

	private getNewAggregatedMap(inputMaps): CodeMap {
		let outputMap: CodeMap = {
			projectName: "Aggregation of following projects: " + this.projectNameArray.join(", "),
			fileName: "Aggregation of following files: " + this.fileNameArray.join(", "),
			nodes: {
				name: "root",
				type: "Folder",
				children: [] as CodeMapNode[],
				attributes: {},
				origin: "Aggregation of following files: " + this.fileNameArray.join(", "),
				path: "/root",
				visible: true
			},
			edges: this.edges,
			blacklist: this.blacklist,
			attributeTypes: {
				nodes: this.attributeTypesNode,
				edges: this.attributeTypesEdge
			}
		}

		for (let inputMap of inputMaps) {
			outputMap.nodes.children.push(this.extractNodeFromMap(inputMap))
		}
		this.aggregateRootAttributes(outputMap)
		return outputMap
	}

	private aggregateRootAttributes(map: CodeMap) {
		map.nodes.children.forEach(child => {
			let attributes = child.attributes
			for (let key in attributes) {
				if (!(key in map.nodes.attributes)) {
					map.nodes.attributes[key] = 0
				}
				map.nodes.attributes[key] += attributes[key]
			}
		})
	}

	private extractNodeFromMap(inputCodeMap: CodeMap): CodeMapNode {
		let outputNode: CodeMapNode = {
			name: inputCodeMap.fileName,
			children: inputCodeMap.nodes.children
		} as CodeMapNode

		if (inputCodeMap.nodes.path) {
			outputNode.path = this.getUpdatedPath(inputCodeMap.fileName, inputCodeMap.nodes.path)
		}

		for (let key in inputCodeMap.nodes) {
			if (!["name", "path", "children"].includes(key)) {
				outputNode[key] = inputCodeMap.nodes[key]
			}
		}
		this.updatePathOfAllChildren(inputCodeMap.fileName, outputNode.children)
		return outputNode
	}

	private updatePathOfAllChildren(fileName, children: CodeMapNode[]) {
		for (let i = 0; i < children.length; i++) {
			if (children[i].path) {
				children[i].path = this.getUpdatedPath(fileName, children[i].path)
			}

			if (children[i].children) {
				this.updatePathOfAllChildren(fileName, children[i].children)
			}
		}
	}

	private getUpdatedPath(fileName, path) {
		const folderArray = path.split("/")
		folderArray.splice(2, 0, fileName)
		return folderArray.join("/")
	}

	private resetVariables() {
		this.projectNameArray = []
		this.fileNameArray = []
		this.edges = []
		this.blacklist = []
		this.attributeTypesEdge = {}
		this.attributeTypesNode = {}
	}
}
