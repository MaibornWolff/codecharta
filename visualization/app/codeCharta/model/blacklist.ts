import { BlacklistItem, BlacklistType } from "../codeCharta.model"
import { addItemToArray, removeItemFromArray } from "../util/reduxHelper"
import ignore from "ignore"
import { CodeMapHelper } from "../util/codeMapHelper"
import _ from "lodash"

export class Blacklist {
	private blacklist: BlacklistItem[] = []

	constructor(blacklist: BlacklistItem[] = []) {
		this.setBlacklist(blacklist)
	}

	public setBlacklist(blacklist: BlacklistItem[]) {
		// TODO revisit clone
		this.blacklist = _.cloneDeep(blacklist)
	}

	public isPathBlacklisted(path: string, type: BlacklistType): boolean {
		if (this.blacklist.length === 0) {
			return false
		}

		const ig = ignore().add(this.blacklist.filter(b => b.type === type).map(ex => CodeMapHelper.transformPath(ex.path)))
		return ig.ignores(CodeMapHelper.transformPath(path))
	}

	public isPathExcluded(path: string): boolean {
		return this.isPathBlacklisted(path, BlacklistType.exclude)
	}

	public isPathFlattened(path: string): boolean {
		return this.isPathBlacklisted(path, BlacklistType.flatten)
	}

	public isPathFlattenedOrExcluded(path: string): boolean {
		return this.isPathExcluded(path) || this.isPathFlattened(path)
	}

	public getItems(): BlacklistItem[] {
		return this.blacklist
	}

	public getItemsByType(type: BlacklistType): BlacklistItem[] {
		return this.blacklist.filter(x => x.type === type)
	}

	public getExcludedItems(): BlacklistItem[] {
		return this.getItemsByType(BlacklistType.exclude)
	}

	public getFlattenedItems(): BlacklistItem[] {
		return this.getItemsByType(BlacklistType.flatten)
	}

	public size(): number {
		return this.blacklist.length
	}

	public sizeByType(type: BlacklistType): number {
		return this.getItemsByType(type).length
	}

	public has(path: string, type: BlacklistType): boolean {
		return this.blacklist.some(x => {
			return x.path === path && x.type === type
		})
	}

	public addBlacklistItem(item: BlacklistItem) {
		this.setBlacklist(addItemToArray(this.blacklist, item))
	}

	public append(blacklist: Blacklist) {
		this.blacklist = this.blacklist.concat(blacklist.getItems())
	}

	public removeBlacklistItem(item: BlacklistItem) {
		this.setBlacklist(removeItemFromArray(this.blacklist, item))
	}
}
