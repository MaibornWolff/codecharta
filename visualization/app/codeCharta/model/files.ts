import { CCFile, FileSelectionState, FileState } from "../codeCharta.model"
const clone = require("rfdc")()

export class Files {
	constructor(private files: FileState[] = []) {}

	public getFiles(): FileState[] {
		return this.files
	}

	public reset() {
		this.files = []
	}

	public resetSelection() {
		this.files = this.files.map(x => {
			return { ...x, selectedAs: FileSelectionState.None }
		})
	}

	public getCCFiles(): CCFile[] {
		return this.files.map(x => x.file)
	}

	public addFile(file: CCFile) {
		this.files = [...this.files, { file: clone(file), selectedAs: FileSelectionState.None }]
	}

	public setFiles(files: Files) {
		this.files = clone(files.getFiles())
	}

	public setSingleByName(fileName: string) {
		this.files = this.files.map(x => {
			if (x.file.fileMeta.fileName === fileName) {
				return { ...x, selectedAs: FileSelectionState.Single }
			} else {
				return { ...x, selectedAs: FileSelectionState.None }
			}
		})
	}

	public setSingle(file: CCFile) {
		this.setSingleByName(file.fileMeta.fileName)
	}

	public setDeltaByNames(referenceFileName: string, comparisonFileName: string) {
		this.files = this.files.map(x => {
			if (x.file.fileMeta.fileName === referenceFileName) {
				return { ...x, selectedAs: FileSelectionState.Reference }
			} else if (x.file.fileMeta.fileName === comparisonFileName) {
				return { ...x, selectedAs: FileSelectionState.Comparison }
			} else {
				return { ...x, selectedAs: FileSelectionState.None }
			}
		})
	}

	public setDelta(reference: CCFile, comparison: CCFile) {
		this.setDeltaByNames(reference.fileMeta.fileName, comparison.fileMeta.fileName)
	}

	public setMultipleByNames(partialFileNames: string[]) {
		this.files = this.files.map(x => {
			if (partialFileNames.includes(x.file.fileMeta.fileName)) {
				return { ...x, selectedAs: FileSelectionState.Partial }
			} else {
				return { ...x, selectedAs: FileSelectionState.None }
			}
		})
	}

	public setMultiple(multipleFiles: CCFile[]) {
		this.setMultipleByNames(multipleFiles.map(x => x.fileMeta.fileName))
	}

	public fileStatesAvailable(): boolean {
		return this.getVisibleFileStates().length > 0
	}

	public getFileByFileName(fileName: string): CCFile {
		const matchingFileState: FileState = this.files.find(x => x.file.fileMeta.fileName == fileName)
		return matchingFileState ? matchingFileState.file : undefined
	}

	public getVisibleFiles(): CCFile[] {
		return this.getVisibleFileStates().map(x => x.file)
	}

	public getVisibleFileStates(): FileState[] {
		return this.files.filter(x => x.selectedAs !== FileSelectionState.None)
	}

	public isSingleState(): boolean {
		return this.files.some(x => x.selectedAs === FileSelectionState.Single)
	}

	public isDeltaState(): boolean {
		return this.files.some(x => x.selectedAs === FileSelectionState.Reference || x.selectedAs === FileSelectionState.Comparison)
	}

	public isPartialState(): boolean {
		return this.files.some(x => x.selectedAs === FileSelectionState.Partial)
	}
}
