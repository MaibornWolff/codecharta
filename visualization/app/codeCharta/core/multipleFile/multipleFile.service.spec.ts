import "./multipleFile.module";
import {NGMock} from "../../../../mocks/ng.mockhelper";
import {AttributeType, BlacklistType, CodeMap, Edge} from "../data/model/CodeMap";

describe("app.codeCharta.core.multiple", function() {

    const file1: CodeMap = {
        fileName: "file1",
        projectName: "Sample Project",
        nodes: {
            name: "root",
            type: "Folder",
            attributes: {"rloc": 170, "functions": 1010, "mcc": 11},
            children: [
                {
                    name: "big leaf",
                    type: "File",
                    attributes: {"rloc": 100, "functions": 10, "mcc": 1},
                    link: "http://www.google.de"
                },
                {
                    name: "Parent Leaf",
                    type: "Folder",
                    attributes: {"rloc": 70, "functions": 1000, "mcc": 10},
                    children: [
                        {
                            name: "other small leaf",
                            type: "File",
                            attributes: {"rloc": 70, "functions": 1000, "mcc": 10}
                        }
                    ]
                }
            ]
        }
    };

    const file2: CodeMap = {
        fileName: "file2",
        projectName: "Sample Project",
        nodes: {
            name: "root",
            type: "Folder",
            attributes: {"rloc" : 260, "functions": 220, "mcc" : 202, "customMetric" : 7},
            children: [
                {
                    name: "big leaf",
                    type: "File",
                    attributes: {"rloc": 200, "functions": 20, "mcc": 2},
                    link: "http://www.google.de"
                },
                {
                    name: "Parent Leaf",
                    type: "Folder",
                    attributes: {"rloc": 60, "functions": 200, "mcc": 200},
                    children: [
                        {
                            name: "small leaf",
                            type: "File",
                            attributes: {"rloc": 60, "functions": 200, "mcc": 200}
                        }
                    ]
                }
            ]
        }
    };

    beforeEach(NGMock.mock.module("app.codeCharta.core.multiple"));

    describe("multipleService", function() {

        it("instance", NGMock.mock.inject(function (multipleFileService) {
            expect(multipleFileService).not.toBe(undefined);
        }));

        it("aggregation of two maps", NGMock.mock.inject(function (multipleFileService) {
            let multiple: CodeMap;
            multiple = multipleFileService.aggregateMaps([file1, file2]);
            expect(multiple).toMatchSnapshot();
        }));

        it("aggregation of four maps", NGMock.mock.inject(function (multipleFileService) {
            let multiple: CodeMap;
            multiple = multipleFileService.aggregateMaps([file1, file2, file1, file2]);
            expect(multiple).toMatchSnapshot();
        }));

        it("aggregation one map", NGMock.mock.inject(function (multipleFileService) {
            let multiple: CodeMap;
            multiple = multipleFileService.aggregateMaps([file1]);
            expect(multiple).toMatchSnapshot();
        }));
        it("aggregate two aggregated maps should aggregate the attributes to root",  NGMock.mock.inject(function (multipleFileService) {
            let multiple: CodeMap;
            multiple = multipleFileService.aggregateMaps([file1, file2]);
            expect(multiple.nodes.attributes.rloc).toBe(430)
            expect(multiple.nodes.attributes.functions).toBe(1230)
            expect(multiple.nodes.attributes.mcc).toBe(213)
            expect(multiple.nodes.attributes.customMetric).toBe(7)
        }));
    });

    describe("edges merge", ()=> {

        const edges1: Edge = [
            {
                fromNodeName: "/root/big leaf",
                toNodeName: "/root/Parent Leaf/small leaf",
                attributes: {
                    pairingRate: 9,
                    avgCommits: 4
                }
            }
        ];

        const edges2: Edge = [
            {
                fromNodeName: "/root/big leaf",
                toNodeName: "/root/Parent Leaf/small leaf",
                attributes: {
                    pairingRate: 89,
                    avgCommits: 34
                }
            }
        ];

        it("aggregation of two maps with empty edges should result in empty edges",NGMock.mock.inject(function(multipleFileService){
            file1.edges = [];
            file2.edges = [];
            let multiple: CodeMap = multipleFileService.aggregateMaps([file1,file2]);
            expect(multiple.edges).toEqual([]);
        }));

        it("aggregation of two maps with given edges should result in summarized edges",NGMock.mock.inject(function(multipleFileService){
            file1.edges = edges1;
            file2.edges = edges2;
            let multiple = multipleFileService.aggregateMaps([file1,file2]);
            expect(multiple.edges).toMatchSnapshot();
        }));

        it("aggregation of one map with edges and other without should result in merged edges",NGMock.mock.inject(function(multipleFileService){
            file1.edges = [edges1[0], edges2[0]];
            file2.edges = null;
            let multiple = multipleFileService.aggregateMaps([file1,file2]);
            expect(multiple.edges).toMatchSnapshot();
        }));

    });


    describe("AttributeTypes merge", ()=> {

        const attribute1 = {
            nodes: {
                ["key1"]: AttributeType.absolute
            },
            edges: {
                ["key2"]: AttributeType.relative
            }
        };

        const attribute2 = {
            nodes: {
                ["key1_2"]: AttributeType.absolute
            },
            edges: {
                ["key2_2"]: AttributeType.relative
            }
        };

        it("aggregation of two maps with different key attributes should result in all keys aggregated map",NGMock.mock.inject(function(multipleFileService){
            file1.attributeTypes = attribute1;
            file2.attributeTypes = attribute2;
            let multiple = multipleFileService.aggregateMaps([file1,file2]);
            expect(multiple.attributeTypes).toMatchSnapshot();
        }));

        it("aggregation of one map with attributes and other without should result in merged with same attributes as the one with them",NGMock.mock.inject(function(multipleFileService){
            file1.attributeTypes = attribute1;
            file2.attributeTypes = {};
            let multiple = multipleFileService.aggregateMaps([file1,file2]);
            expect(multiple.attributeTypes).toMatchSnapshot();
        }));

    });

    describe("Blacklist merge", ()=> {

        const blacklist1 = [
            {
                path: "/root/node/equalPath",
                type: BlacklistType.exclude
            },
            {
                path: "/another/third/path",
                type: BlacklistType.hide
            }
        ];

        const blacklist2 = [
            {
                path: "/root/node/equalPath",
                type: BlacklistType.exclude
            },
            {
                path: "*another/path",
                type: BlacklistType.hide
            }
        ];

        it("aggregation of two maps with different key attributes should result in all keys aggregated map",NGMock.mock.inject(function(multipleFileService){
            file1.blacklist = blacklist1;
            file2.blacklist = blacklist2;
            let multiple = multipleFileService.aggregateMaps([file1,file2]);
            expect(multiple.blacklist).toMatchSnapshot();
        }));
    });
});