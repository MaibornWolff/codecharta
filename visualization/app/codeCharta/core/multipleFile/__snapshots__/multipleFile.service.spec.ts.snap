// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app.codeCharta.core.multiple AttributeTypes merge aggregation of one map with attributes and other without should result in merged with same attributes as the one with them 1`] = `
Object {
  "edges": Object {
    "key2": 1,
  },
  "nodes": Object {
    "key1": 0,
  },
}
`;

exports[`app.codeCharta.core.multiple AttributeTypes merge aggregation of two maps with different key attributes should result in all keys aggregated map 1`] = `
Object {
  "edges": Object {
    "key2": 1,
    "key2_2": 1,
  },
  "nodes": Object {
    "key1": 0,
    "key1_2": 0,
  },
}
`;

exports[`app.codeCharta.core.multiple Blacklist merge aggregation of two maps with different key attributes should result in all keys aggregated map 1`] = `
Array [
  Object {
    "path": "/root/file1/node/equalPath",
    "type": "exclude",
  },
  Object {
    "path": "/another/third/path",
    "type": "hide",
  },
  Object {
    "path": "/root/file2/node/equalPath",
    "type": "exclude",
  },
  Object {
    "path": "*another/path",
    "type": "hide",
  },
]
`;

exports[`app.codeCharta.core.multiple edges merge aggregation of one map with edges and other without should result in merged edges 1`] = `
Array [
  Object {
    "attributes": Object {
      "avgCommits": 4,
      "pairingRate": 9,
    },
    "fromNodeName": "/root/file1/big leaf",
    "toNodeName": "/root/file1/Parent Leaf/small leaf",
    "visible": undefined,
  },
]
`;

exports[`app.codeCharta.core.multiple edges merge aggregation of two maps with given edges should result in summarized edges 1`] = `
Array [
  Object {
    "attributes": Object {
      "avgCommits": 4,
      "pairingRate": 9,
    },
    "fromNodeName": "/root/file1/big leaf",
    "toNodeName": "/root/file1/Parent Leaf/small leaf",
    "visible": undefined,
  },
  Object {
    "attributes": Object {
      "avgCommits": 34,
      "pairingRate": 89,
    },
    "fromNodeName": "/root/file2/big leaf",
    "toNodeName": "/root/file2/Parent Leaf/small leaf",
    "visible": undefined,
  },
]
`;

exports[`app.codeCharta.core.multiple multipleService aggregation of four maps 1`] = `
Object {
  "attributeTypes": Object {
    "edges": Object {},
    "nodes": Object {},
  },
  "blacklist": Array [],
  "edges": Array [],
  "fileName": "Aggregation of following files: file1, file2, file1, file2",
  "nodes": Object {
    "attributes": Object {
      "customMetric": 14,
      "functions": 2460,
      "mcc": 426,
      "rloc": 860,
    },
    "children": Array [
      Object {
        "attributes": Object {
          "functions": 1010,
          "mcc": 11,
          "rloc": 170,
        },
        "children": Array [
          Object {
            "attributes": Object {
              "functions": 10,
              "mcc": 1,
              "rloc": 100,
            },
            "link": "http://www.google.de",
            "name": "big leaf",
            "type": "File",
          },
          Object {
            "attributes": Object {
              "functions": 1000,
              "mcc": 10,
              "rloc": 70,
            },
            "children": Array [
              Object {
                "attributes": Object {
                  "functions": 1000,
                  "mcc": 10,
                  "rloc": 70,
                },
                "name": "other small leaf",
                "type": "File",
              },
            ],
            "name": "Parent Leaf",
            "type": "Folder",
          },
        ],
        "name": "file1",
        "type": "Folder",
      },
      Object {
        "attributes": Object {
          "customMetric": 7,
          "functions": 220,
          "mcc": 202,
          "rloc": 260,
        },
        "children": Array [
          Object {
            "attributes": Object {
              "functions": 20,
              "mcc": 2,
              "rloc": 200,
            },
            "link": "http://www.google.de",
            "name": "big leaf",
            "type": "File",
          },
          Object {
            "attributes": Object {
              "functions": 200,
              "mcc": 200,
              "rloc": 60,
            },
            "children": Array [
              Object {
                "attributes": Object {
                  "functions": 200,
                  "mcc": 200,
                  "rloc": 60,
                },
                "name": "small leaf",
                "type": "File",
              },
            ],
            "name": "Parent Leaf",
            "type": "Folder",
          },
        ],
        "name": "file2",
        "type": "Folder",
      },
      Object {
        "attributes": Object {
          "functions": 1010,
          "mcc": 11,
          "rloc": 170,
        },
        "children": Array [
          Object {
            "attributes": Object {
              "functions": 10,
              "mcc": 1,
              "rloc": 100,
            },
            "link": "http://www.google.de",
            "name": "big leaf",
            "type": "File",
          },
          Object {
            "attributes": Object {
              "functions": 1000,
              "mcc": 10,
              "rloc": 70,
            },
            "children": Array [
              Object {
                "attributes": Object {
                  "functions": 1000,
                  "mcc": 10,
                  "rloc": 70,
                },
                "name": "other small leaf",
                "type": "File",
              },
            ],
            "name": "Parent Leaf",
            "type": "Folder",
          },
        ],
        "name": "file1",
        "type": "Folder",
      },
      Object {
        "attributes": Object {
          "customMetric": 7,
          "functions": 220,
          "mcc": 202,
          "rloc": 260,
        },
        "children": Array [
          Object {
            "attributes": Object {
              "functions": 20,
              "mcc": 2,
              "rloc": 200,
            },
            "link": "http://www.google.de",
            "name": "big leaf",
            "type": "File",
          },
          Object {
            "attributes": Object {
              "functions": 200,
              "mcc": 200,
              "rloc": 60,
            },
            "children": Array [
              Object {
                "attributes": Object {
                  "functions": 200,
                  "mcc": 200,
                  "rloc": 60,
                },
                "name": "small leaf",
                "type": "File",
              },
            ],
            "name": "Parent Leaf",
            "type": "Folder",
          },
        ],
        "name": "file2",
        "type": "Folder",
      },
    ],
    "name": "root",
    "origin": "Aggregation of following files: file1, file2, file1, file2",
    "path": "/root",
    "type": "Folder",
    "visible": true,
  },
  "projectName": "Aggregation of following projects: Sample Project, Sample Project, Sample Project, Sample Project",
}
`;

exports[`app.codeCharta.core.multiple multipleService aggregation of two maps 1`] = `
Object {
  "attributeTypes": Object {
    "edges": Object {},
    "nodes": Object {},
  },
  "blacklist": Array [],
  "edges": Array [],
  "fileName": "Aggregation of following files: file1, file2",
  "nodes": Object {
    "attributes": Object {
      "customMetric": 7,
      "functions": 1230,
      "mcc": 213,
      "rloc": 430,
    },
    "children": Array [
      Object {
        "attributes": Object {
          "functions": 1010,
          "mcc": 11,
          "rloc": 170,
        },
        "children": Array [
          Object {
            "attributes": Object {
              "functions": 10,
              "mcc": 1,
              "rloc": 100,
            },
            "link": "http://www.google.de",
            "name": "big leaf",
            "type": "File",
          },
          Object {
            "attributes": Object {
              "functions": 1000,
              "mcc": 10,
              "rloc": 70,
            },
            "children": Array [
              Object {
                "attributes": Object {
                  "functions": 1000,
                  "mcc": 10,
                  "rloc": 70,
                },
                "name": "other small leaf",
                "type": "File",
              },
            ],
            "name": "Parent Leaf",
            "type": "Folder",
          },
        ],
        "name": "file1",
        "type": "Folder",
      },
      Object {
        "attributes": Object {
          "customMetric": 7,
          "functions": 220,
          "mcc": 202,
          "rloc": 260,
        },
        "children": Array [
          Object {
            "attributes": Object {
              "functions": 20,
              "mcc": 2,
              "rloc": 200,
            },
            "link": "http://www.google.de",
            "name": "big leaf",
            "type": "File",
          },
          Object {
            "attributes": Object {
              "functions": 200,
              "mcc": 200,
              "rloc": 60,
            },
            "children": Array [
              Object {
                "attributes": Object {
                  "functions": 200,
                  "mcc": 200,
                  "rloc": 60,
                },
                "name": "small leaf",
                "type": "File",
              },
            ],
            "name": "Parent Leaf",
            "type": "Folder",
          },
        ],
        "name": "file2",
        "type": "Folder",
      },
    ],
    "name": "root",
    "origin": "Aggregation of following files: file1, file2",
    "path": "/root",
    "type": "Folder",
    "visible": true,
  },
  "projectName": "Aggregation of following projects: Sample Project, Sample Project",
}
`;

exports[`app.codeCharta.core.multiple multipleService aggregation one map 1`] = `
Object {
  "fileName": "file1",
  "nodes": Object {
    "attributes": Object {
      "functions": 1010,
      "mcc": 11,
      "rloc": 170,
    },
    "children": Array [
      Object {
        "attributes": Object {
          "functions": 10,
          "mcc": 1,
          "rloc": 100,
        },
        "link": "http://www.google.de",
        "name": "big leaf",
        "type": "File",
      },
      Object {
        "attributes": Object {
          "functions": 1000,
          "mcc": 10,
          "rloc": 70,
        },
        "children": Array [
          Object {
            "attributes": Object {
              "functions": 1000,
              "mcc": 10,
              "rloc": 70,
            },
            "name": "other small leaf",
            "type": "File",
          },
        ],
        "name": "Parent Leaf",
        "type": "Folder",
      },
    ],
    "name": "root",
    "type": "Folder",
  },
  "projectName": "Sample Project",
}
`;
